packages/cli/src/UserManagement/PermissionChecker.ts

```diff
@@ -1,68 +1,76 @@
 import { Service } from 'typedi';
 import type { INode, Workflow } from 'n8n-workflow';
-import { CredentialAccessError, NodeOperationError, WorkflowOperationError } from 'n8n-workflow';
+import {
+	ApplicationError,
+	CredentialAccessError,
+	NodeOperationError,
+	WorkflowOperationError,
+} from 'n8n-workflow';
 
 import config from '@/config';
 import { License } from '@/License';
 import { OwnershipService } from '@/services/ownership.service';
 import { UserRepository } from '@db/repositories/user.repository';
 import { SharedCredentialsRepository } from '@db/repositories/sharedCredentials.repository';
-import { SharedWorkflowRepository } from '@db/repositories/sharedWorkflow.repository';
+import { ProjectService } from '@/services/project.service';
+import { RoleService } from '@/services/role.service';
 
 @Service()
 export class PermissionChecker {
 	constructor(
 		private readonly userRepository: UserRepository,
 		private readonly sharedCredentialsRepository: SharedCredentialsRepository,
-		private readonly sharedWorkflowRepository: SharedWorkflowRepository,
 		private readonly ownershipService: OwnershipService,
 		private readonly license: License,
+		private readonly projectService: ProjectService,
+		private readonly roleService: RoleService,
 	) {}
 
 	/**
 	 * Check if a user is permitted to execute a workflow.
+	 *
+	 * Allow:
+	 * - if user has global scope `workflow:execute`, OR
+	 * - if user has project scope `workflow:execute` in any of their roles
+	 * in any of the projects where the workflow is accessible, AND if every
+	 * credential used by nodes in the workflow is accessible by any of the
+	 * projects where the workflow is accessible.
 	 */
 	async check(workflowId: string, userId: string, nodes: INode[]) {
-		// allow if no nodes in this workflow use creds
-
-		const credIdsToNodes = this.mapCredIdsToNodes(nodes);
-
-		const workflowCredIds = Object.keys(credIdsToNodes);
-
-		if (workflowCredIds.length === 0) return;
-
-		// allow if requesting user is instance owner
-
-		const user = await this.userRepository.findOneOrFail({
-			where: { id: userId },
-		});
+		const user = await this.userRepository.findOneByOrFail({ id: userId });
 
 		if (user.hasGlobalScope('workflow:execute')) return;
 
-		const isSharingEnabled = this.license.isSharingEnabled();
-
-		// allow if all creds used in this workflow are a subset of
-		// all creds accessible to users who have access to this workflow
+		const { roles, projectIds } = await this.projectService.findRolesAndProjects(
```

Comment:
Would it make sense to change the scope checks to the middleware instead?

I'm asking because the way it's currently implemented, we might have some unexpected downsides in the future.

Say I'm a regular `Member` in an instance, and create a number of workflows. Then at some point my role is changed to `Viewer` - I would expect to no longer be able to manually execute workflows, but the production workflows I own (i.e. are part of my own personal project) should still continue to run, I think.

This would break with the current code.

Reply:
Also I'd say this function should be used to check for credential permissions alone, and not user permissions too. This would be violating the Single Responsibility Principle.